4단계 - 볼링 점수판(n명)

nooose opened the pull request:
준영님, 마지막 미션입니다!
3단계 제출하고 리뷰전 까지 4단계를 미리 진행하고 있어서 rebase 하고 PR 올리느라 커밋내역이 겹치네요 😅

#912 피드백도 반영했습니다.
리뷰 잘 부탁드립니다!

nooose committed:
refactor: 불필요한 오버라이드 메서드 제거

#899 피드백 반영
동일한 로직이므로 제거한다.

nooose committed:
refactor: 도메인 메서드 명 수정

pitch 보다 더 자연스러운 bowl 로 변경한다.

nooose committed:
feat: Pin 도메인 기능 구현

핀이 모두 쓰러졌는지 알 수 있고, 핀 끼리 합 계산이 가능하도록 구현했다.

nooose committed:
refactor: 상태 패턴으로 변경

지저분했던 if 분기를 제거하는 동시에 상태 전환 로직이 보기 좋아졌다.

nooose committed:
refactor: State 에서 점수로 반환할 값을 객체로 포장

nooose committed:
feat: 점수 계산 기능 구현

nooose committed:
feat: 점수 출력 뷰 기능 구현

nooose committed:
refactor: 객체와 메서드 이름 변경

- V2 prefix 제거
리팩토링 과정 중 동시에 존재했을 때 붙인 이름으로 지금은 제거해도 괜찮다.
- Pitch -> Bowl
도메인 메서드 용어가 수정됨에 따라 같이 변경했다.

nooose committed:
refactor: Override 어노테이션 추가

nooose committed:
refactor: 출력 역할 분리

nooose committed:
refactor: 불필요한 try-catch 제거

추가로 리스트 편의 메서드 구현

nooose committed:
refactor: 도메인이 뷰를 의존하지 않도록 변경

nooose committed:
test: assertAll 사용

nooose committed:
refactor: 불필요하게 열려있는 접근제어자 수정

그리고, 상수가 가능한 값은 치환시켰다.

nooose committed:
fix: 마지막 프레임 점수 미출력 문제 수정

마지막 프레임이 MISS 인 경우 점수가 제대로 표시되지 않았다.

nooose committed:
test: 부족했던 테스트 코드 추가

빠진 테스트 코드를 추가했다.

nooose committed:
fix: 마지막 프레임에서 사용자가 잘 못 입력했을 때 기호가 제대로 표시되지 않은 문제 수정

nooose committed:
feat: 사용자 입력이 잘못된 경우 정확한 메시지를 출력하도록 변경

nooose committed:
refactor: 사용하지 않는 메서드 제거

nooose committed:
refactor: Controller 부분 메서드 분리

depth 를 줄이기 위해 분리 시켰다.

nooose committed:
docs: 기능 요구사항 업데이트

요구사항이 완료되어 리스트를 추가함

nooose committed:
refactor: score 필드 제거

state를 통해서 얻을 수 있다.

nooose committed:
refactor: 공통 부분 추상화

그리고 메서드 순서를 일관되게 변경했다.

nooose committed:
fix: 점수를 제대로 출력하지 못하는 문제 수정

state 필드를 통해 얻을 수 있다고 생각해 score 필드를 지웠으나,
누적된 점수를 저장하지 못해 점수를 구하지 못했다.

nooose committed:
refactor: 클래스 타입을 식별하는 메서드 제거

nooose committed:
refactor: 볼링 투구를 플레이어가 진행하도록 변경

nooose committed:
feat: 투구 시 플레이어의 이름을 출력하도록 변경

nooose committed:
refactor: 출력 로직 개선

출력되는 점수 또는 기호를 이어붙이지 않고 프레임 템플릿 포맷에 맞게 출력하도록 변경

nooose committed:
fix: 마지막 프레임 기호가 잘못 출력되는 문제 수정

nooose committed:
fix: 일반 프레임 Miss 인 경우 점수 계산이 제대로 되지 않는 문제 수정

nooose committed:
feat: Player 일급 컬렉션 구현

nooose committed:
feat: 여러 플레이어에 대해 입출력이 가능하도록 구현

nooose committed:
refactor: 메서드 분리

nooose committed:
refactor: 변수명 변경

이전 리팩토링 때 사용했던 V2 suffix 를 삭제한다.

nooose committed:
refactor: 객체를 재사용할 수 있도록 수정

nooose committed:
Merge branch 'nooose' into step4

nooose committed:
feat: 중복된 플레이어 이름 방지 기능 구현

nooose committed:
refactor: null 체크 표현 리팩토링

#912 피드백 반영

nooose committed:
refactor: State 추상화를 통해 isFinished 중복 메서드 제거

nooose committed:
docs: 추가 기능구현에 맞게 문서 업데이트

nooose reviewed (commented):
-

nooose committed:
refactor: 불필요한 static 키워드 제거

nooose committed:
refactor: 컨트롤러 내 private 메서드 변경

가독성을 위해 변경함

nooose committed:
test: Player 일급 컬렉션 테스트 코드 작성

nooose committed:
test: assertJ 메서드로 변경

nooose committed:
refactor: 정의된 상수값을 사용하도록 변경

nooose committed:
refactor: 변수명 수정

Score 가 score 를 반환하는 경우 헷갈릴 수 있어
getter 메서드와 변수명을 수정했다.

lxxjn0 commented on src/main/java/bowling/controller/BowlingController.java:
```diff
@@ -1,54 +1,71 @@
-        String playerName = InputView.inputPlayerName();
+    private Players getPlayers() {
+        int numberOfPlayers = InputView.inputNumberOfPlayers();
+        List<Player> players = IntStream.rangeClosed(1, numberOfPlayers)
```
만약 numberOfPlayers의 크기만큼 반복을 하고 싶은거라면 `range(0, nubmerOfPlayers)` 같은 로직도 괜찮을 것 같아요!! for문이 index 0으로 시작하는 걸 기본 로직으로 가지고 있다보니 `range(0, nubmerOfPlayers)` 와 같은 코드가 좀 더 이해가 잘 가지 않을까?? 싶은 생각이 들어 남겨봤습니다!! ㅎㅎㅎ

이건 중요한 피드백은 아니고 저의 개인적인 생각이라 참고만 해주세요 :)

nooose replied:
<img width="366" alt="image" src="https://user-images.githubusercontent.com/36617783/201523535-b7287ddb-f7d9-452d-ba3f-624886c41a42.png">
위 사진 처럼 index가 1 부터 출력시키기 위해서 
`IntStream.rangeClosed(1, numberOfPlayers)`  이렇게 사용하고 있는데요

답변 주신대로 `IntStream.range(0, numberOfPlayers)` 로 반복을 돌리고 index 를 따로 `+1` 증가시켜야 하는 걸까요?
제가 이 부분은 잘 이해를 못한 것 같습니다.😭

nooose replied:
준영님의 의견은 사이즈만큼 반복시킬 때
range 메서드로 0 부터 Size 크기만큼 반복을 돌리는 것이 좀 더 이해하기 쉽다 라는 의미일까요?

저는 불필요하게 +1 을 할까봐 rangeClosed를 사용했었는데 
무조건 반복된 사이즈만큼 반복시키는 경우라면 range(0, size)로 범위를 확실히 보여주는 것도 괜찮은 방법이겠네요

lxxjn0 replied:
제가 출력문에서 해당 인덱스를 사용하는 상황을 따로 인지하진 않았었네요 ㅠㅠ 그렇다면 작성해주신 로직도 충분히 좋다고 생각해요! 또는 제가 작성한 로직에서 +1을 하는 방법도 있겠지만 지금처럼 출력에서 활용을 한다면 준혁님 로직도 좋다고 생각합니다 :)

lxxjn0 commented on src/main/java/bowling/controller/BowlingController.java:
```diff
@@ -1,54 +1,71 @@
 
-    private Player getPlayer() {
-        String playerName = InputView.inputPlayerName();
+    private Players getPlayers() {
```
이 부분도 개인적인 의견이지만 get을 prefix로 쓰는 경우 보통 getter로 인식이 되다보니 저는 getter가 아닌 메서드에서는 get을 사용하는 걸 지양하는 편이에요. `generatePlayers()`나 `createPlayers()` 등 다른 이름을 사용해보는 것도 좋을 것 같아요!! 👍🏼

lxxjn0 commented on src/main/java/bowling/domain/Score.java:
```diff
@@ -1,24 +1,24 @@
 
 public class Score {
-    private final int score;
+    private final int value;
```
이 부분은 전 개인적으로 이전에 score가 더 좋은 것 같아요!! value라는 명은 특정 도메인의 필드를 나타내기에는 너무 범용적이고 모호한 단어가 아닐까 싶어요!! 중요한 부분은 아니기에 한 번 고민해보면 좋을 것 같아요!! 

`score.getScore()`는 Score 객체가 관리하는 score라는 값을 반환하는구나!! 라는 것을 알겠는데 `score.getValue()`는 어떠한 value인지 코드를 보지 않으면 명확히 식별이 안되는 것 같아서 아쉬웠어요

nooose replied:
앗 그렇군요.
score 에서 또 getScore() 를 하다보니 스코어에서 또 스코어를 꺼낸다? 라는 느낌이 좀 이상해서 변경했었습니다.
이런 경우에는 변수명을 score로 두고 클래스명을 바꾸도록 해볼게요

nooose replied:
Name 클래스와 Score 클래스에서도 value 라는 필드명을 사용했었는데요

리뷰어님 말씀대로 생각도해보니까 특정 객체가 관리하는 값을 반환한다는 의미가 있네요.
중복적인 표현같았었는데 이렇게도 생각할 수 있네요...

위 코멘트에서 클래스 명도 변경해보려고 했는데 FrameScore 는 Frame 이 중복적으로 들어가서 별로인 것 같구..
Point 도 생각해봤지만 Score 만 계속봐서 그런지 Score 가 더 자연스러운 것 같네요. 😅

변경하도록 하겠습니다.

lxxjn0 commented on src/main/java/bowling/domain/player/Name.java:
```diff
@@ -18,7 +18,7 @@ public Name(String value) {
 
     private static void validateNullOrBlank(String value) {
-        if (Objects.isNull(value) || value.isBlank()) {
+        if (value == null || value.isBlank()) {
```
굿!! 👍🏼

lxxjn0 commented on src/main/java/bowling/controller/BowlingController.java:
```diff
@@ -1,54 +1,71 @@
+        }
+    }
+
+    private void inputBowlNumberAndPrintResult(Player player, Players players) {
+        if (player.canBowl()) {
+            inputBowlNumber(player);
+            ResultView.printFrameResults(players);
+        }
     }
```
여기에 있는 출력 로직을 상단 메서드로 옮기는 건 어떨까요?? 그렇다면 Players 파라미터를 받지 않아도 될 것 같아요!! bowl이라는 private 메서드에서 출력까지 담당한다라는 느낌으로 봐도 괜찮을 것 같아요!!

nooose replied:
그렇네요! 감사합니다.

lxxjn0 commented on src/main/java/bowling/controller/BowlingController.java:
```diff
@@ -1,54 +1,71 @@
 
-        frames.addFrame();
+    private void bowl(Players players) {
+        for (Player player : players.getPlayers()) {
```
일급 컬렉션으로 감싼 객체를 다시 getter로 가져와서 로직을 진행하는 것이 '일급 컬렉션으로 사용하는 의미가 있는걸까?'에 대한 고민이 들긴 하네요.. 그런데 입출력 로직이 관여되어 있다보니 어쩔 수 없을 것 같아요 ㅠㅠ 콘솔 프로그램이라서 가지는 한계라는 생각은 들어서 아쉽지만 해당 부분을 개선하기는 어려움이 있을 것 같네요 ㅠㅠ

nooose replied:
넵, 말씀주신대로 저도 공감하는 부분입니다..
드라마틱하게 느끼는 장점은 없지만
모든 플레이어들의 게임 종료여부 상태 체크, 중복 회원 체크만으로도 컨트롤러 내 코드가 많이 줄어들어 좋은 것 같습니다.

lxxjn0 reviewed (changes requested):
준혁님, 미션 진행하시느라 수고하셨습니다!!

크게 로직적인 부분에서 변경 사항은 없어서 크게 피드백을 드릴만한 부분이 없네요!! 간단한 부분에 대해서만 피드백 남겨두었으니 한 번 확인해보시고 반영 부탁드릴게요 :)

nooose committed:
refactor: 출력 로직 위치 변경

Players 파라미터를 전달하지 않아도 처리가 가능하다.

nooose committed:
refactor: Controller 내 메서드 위치 및 표현 개선

#952 피드백 반영
- players 파라미터를 불필요하게 전달하는 부분 제거
- getter 메서드가 아닌 메서드에서 get prefix 를 제거하고 create 로 변경함
- 반복 사이즈를 명확하게 표현하기 위해 range 메서드를 사용

nooose committed:
refactor: 애매한 필드 명 변경

#952 피드백 반영

nooose commented on src/main/java/bowling/domain/player/Players.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.HashSet;
+import java.util.List;
+
+public class Players {
+
+    private final List<Player> values;
+
```
일급 컬렉션의 경우에도 동일할까요?

저는 `getXXXs` 와 같이 값을 꺼낼 때 컬렉션이기 때문에 values 라는 값을 꺼내면 
해당 일급 컬렉션이 담고있는 객체들을 반환한다는 게 명확해 보일 수 있다고 생각합니다.
Enum 의 values() 처럼요!

리뷰어님의 의견이 궁금합니다.

lxxjn0 replied:
이 부분은 개인적인 의견의 차이일 것 같아요!! :)

enum의 경우 자바 자체에서 제공해주는 특별한 클래스 타입이고 values는 그러한 특정한 클래스에서 제공해주는 default 메서드이다보니 모든 사람들이 합의가 된 코드라고 생각해요. 

그러나 우리가 구현하는 Players와 같은 일급 컬렉션은 실제 내부 구현을 확인하지 않는 이상은 일급 컬렉션이라는 사실을 모를 것 같아요. 실제 코드의 내부를 봐야지 '아, 일급 컬렉션이라 필드가 컬렉션 타입으로 하나만 가지고 있으니깐 values라는 변수명을 준거구나'라고 이해할 수 있을 것 같아요!! 

물론 사실 내부 코드를 아예 안본다는 가정도 말은 안되고 일급 컬렉션에 한해서는 values라는 필드를 쓰는 것도 괜찮다고는 생각해요. 그러나 players로 표현하는 방법이 보기에 어색하다? 말고는 values에 비해 뚜렷하게 가지는 단점이 없다고 생각해서 저라면 players로 필드 명을 사용하고 getter도 동일한 형식으로 가지 않을까 싶어요. 

실제 변수명도 변수가 가지는 의미를 담아서 만들어야 하는데 Player 다수를 담는 컬렉션을 표현하기 위한 변수이므로 players도 적절한 변수명이라 생각하기도 하구요!!

그러나 이런 변수 네이밍은 어느정도 합의가 된다면 괜찮다고 생각해서 개인의 생각에 따라서 팀에서 이해가는 범위 안에 용어를 사용하면 충분할 것 같아요!!

nooose replied:
상세한 답변 정말 감사합니다!

lxxjn0 reviewed (approved):
준혁님, 미션 진행하시느라 정말 수고 많으셨습니다 👍🏼 

추가적인 피드백을 드릴만한 부분이 없는 것 같아서 이번 미션은 여기서 마무리하도록 할게요!! 고생 많으셨어요 :)
